ddtrace/tracer/sampler_test.go

```diff
@@ -591,3 +751,128 @@ func BenchmarkRulesSampler(b *testing.B) {
+	} {
+		t.Run(fmt.Sprintf("%d", i), func(t *testing.T) {
+			rg, _ := globMatch(tt.pattern)
+			if tt.shouldMatch {
+				assert.Regexp(t, rg, tt.input)
+			} else {
+				assert.NotRegexp(t, rg, tt.input)
+			}
+		})
+	}
+}
+
+func TestSamplingRuleMarshall(t *testing.T) {
+	for _, tt := range []struct {
+		in  SamplingRule
+		out string
+	}{
+		{SamplingRule{nil, nil, 0, 0, 0, "srv", "ops", nil},
+			`{"service":"srv","name":"ops","sample_rate":0,"type":0}`},
+		{SamplingRule{regexp.MustCompile("srv.[0-9]+]"), nil, 0, 0, 0, "srv", "ops", nil},
+			`{"service":"srv","name":"ops","sample_rate":0,"type":0}`},
+		{SamplingRule{regexp.MustCompile("srv.*"), regexp.MustCompile("ops.[0-9]+]"), 0, 0, 0, "", "", nil},
+			`{"service":"srv.*","name":"ops.[0-9]+]","sample_rate":0,"type":0}`},
+		{SamplingRule{regexp.MustCompile("srv.[0-9]+]"), regexp.MustCompile("ops.[0-9]+]"), 0.55, 0, 0, "", "", nil},
+			`{"service":"srv.[0-9]+]","name":"ops.[0-9]+]","sample_rate":0.55,"type":0}`},
+		{SamplingRule{regexp.MustCompile("srv.[0-9]+]"), regexp.MustCompile("ops.[0-9]+]"), 0.55, 0, 1, "", "", nil},
+			`{"service":"srv.[0-9]+]","name":"ops.[0-9]+]","sample_rate":0.55,"type":1}`},
+		{SamplingRule{regexp.MustCompile("srv.[0-9]+]"), regexp.MustCompile("ops.[0-9]+]"), 0.55, 1000, 1, "", "", nil},
+			`{"service":"srv.[0-9]+]","name":"ops.[0-9]+]","sample_rate":0.55,"type":1,"max_per_second":1000}`},
+	} {
+		m, err := tt.in.MarshalJSON()
+		assert.Nil(t, err)
+		assert.Equal(t, tt.out, string(m))
+	}
+}
+func BenchmarkGlobMatchSpan(b *testing.B) {
+	var spans []*span
+	for i := 0; i < 1000; i++ {
+		spans = append(spans, newSpan("name.ops.date", "srv.name.ops.date", "", 0, 0, 0))
+	}
+
+	b.Run("no-regex", func(b *testing.B) {
+		os.Setenv("DD_SPAN_SAMPLING_RULES", `[{"service": "srv.name.ops.date", name:"name.ops.date?", sample_rate": 0.234}]`)
+		os.Unsetenv("DD_SPAN_SAMPLING_RULES")
+		rules, err := samplingRulesFromEnv()
+		assert.Nil(b, err)
+		rs := newSingleSpanRulesSampler(rules)
+		b.ResetTimer()
+		for n := 0; n < b.N; n++ {
+			for _, span := range spans {
+				rs.apply(span)
+			}
+		}
+	})
+
+	b.Run("glob-match-?", func(b *testing.B) {
+		os.Setenv("DD_SPAN_SAMPLING_RULES", `[{"service": "srv?name?ops?date", name:"name*ops*date*", sample_rate": 0.234}]`)
+		os.Unsetenv("DD_SPAN_SAMPLING_RULES")
+		rules, err := samplingRulesFromEnv()
+		assert.Nil(b, err)
+		rs := newSingleSpanRulesSampler(rules)
+		b.ResetTimer()
+		for n := 0; n < b.N; n++ {
+			for _, span := range spans {
+				rs.apply(span)
+			}
+		}
+	})
+
+	b.Run("glob-match-*", func(b *testing.B) {
+		os.Setenv("DD_SPAN_SAMPLING_RULES", `[{"service": "srv*name*ops*date", name:"name?ops?date?", sample_rate": 0.234}]`)
+		os.Unsetenv("DD_SPAN_SAMPLING_RULES")
+
+		rules, err := samplingRulesFromEnv()
+		assert.Nil(b, err)
+		rs := newSingleSpanRulesSampler(rules)
+
+		b.ResetTimer()
+		for n := 0; n < b.N; n++ {
+			for _, span := range spans {
+				rs.apply(span)
+			}
+		}
+	})
```

Comment:
Consider adding one of the nasty ones: https://github.com/DataDog/tag-matching-sampling-rules/blob/f1e18c73594e047f65effae7c07f26d5eb2438f5/glob.mjs#L74-L75